### PR TITLE
Add example using the ws281x-rpi driver

### PR DESCRIPTION
**Pull request by nathansamson**
Add an example for Raspberry Pi using an alternative driver that allows full use of the std library and OS.

This is both intended as an example for people as a bit of (self) advertisement for this alternative driver.

### CODE REVIEW

**Comment by sajattack** [comment]
Hmmm, why not just use linux-embedded-hal::spi and ws2812_spi?

**Comment by nathansamson** [comment]
In all honesty... Because I could not get it to work :)  (note I used rppal but I think I would have the same issue).
Note that in the meantime I think I figured out what I probably did wrong and I could make this work.

That being said I still think its worth including this driver as reference + example as this driver is
A) Based on another driver that seems pretty popular (I found it by googling tutorials about connecting ws2812 to my raspberry pi). This means that problems people might have has quite a big community around it to solve issues (which sometimes you might need to convert to rust, but in some cases is just a OS configuration issue).

B) This driver gives 3 different modes (PCM, PWM, SPI) and gives users a bit more freedom. All modes have their own limitations. Admittedly SPI is the easier interface. (although previously mentioned tutorial uses PWM by default).


I hope this convinces you that this example is still worth it. If not I am probably going to create one based on either rppal and/or linux embedded hal as you suggested as I feel this might help other people as well figuring out how to make it work on a raspberry pi.

**Comment by david-sawatzke** [comment]
I think this is a worthwhile addition and (after some testing) would like to merge it. The driver you've written also looks pretty good on a cursory overview, thank you.

Since this isn't using a `smart-leds` crate, please add a note to the readme that this example uses an external crate and is thus supported by the author (well, you). 

Adding a backend to (better) support non-baremetal targets to ws2812-spi is something I'd like to see in the future, but ws281x-rpi is more flexible on the rpi and is available right now.

File: raspberry-pi-ws281x-rpi-examples/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+#![no_std]
```
Comment (david-sawatzke) [inline raspberry-pi-ws281x-rpi-examples/src/lib.rs:1]: I don't think this file needs to exist

**Review by david-sawatzke — CHANGES_REQUESTED** [review]
I think many people would want to build it on an x86 host instead of on the rpi (especially for the first, zero and zero w). A `.cargo/config` with the correct build target would help.

File: raspberry-pi-ws281x-rpi-examples/src/lib.rs (hunk 1 above)
Comment (david-sawatzke) [reply raspberry-pi-ws281x-rpi-examples/src/lib.rs:1]: Sorry, my bad

**Comment by david-sawatzke** [comment]
I'm giving up on crosscompiling the example: 
- Using arm-unknown-linux-gnueabihf on arch doesn't work
- Cross doesn't work
- Using zig also fails

Damm you, C!

**Comment by david-sawatzke** [comment]
Seems to work pretty good on the raspberry pi one. After you change the README to note that this uses a third-party crate, I think this PR can be merged.

**Comment by david-sawatzke** [comment]
FYI: I've made https://github.com/smart-leds-rs/ws2812-spi-rs/pull/27 as a pure-rust solution. Less tested and more limited pin-wise, may interest you.

That one can be crosscompiled using:
```
[build]
target = "arm-unknown-linux-musleabihf"
# rustup target add arm-unknown-linux-musleabihf


[target.arm-unknown-linux-musleabihf]
linker = "rust-lld"
```
But I'm pretty sure that would break ws281x-rpi.

Gonna have to look somewhat deeper in the cargo documentation or create a separate crate for the example.

It's not a blocker for this PR

**Comment by nathansamson** [comment]
Updated the readme to clarify third-party status of driver used. (Just adding one word, if you think more is needed I'll add this :))

About cross-compile:
I did not mention any cross compile instructions as I figured out it would be too hard.
Also presumably users are using this with a Pi OS which should be more than capable enough of compiling rust applications in a decent amount of time and is less needed (eg when using the other rpi examples where it is being used in the boot sector you might not have an ARM device on hand, but in this use case people obviously do)

Thanks for working on the hosted variant, I might give that a try as well

**Pull request merged**